[FIRRTL] Add alternative constraint-based InferWidths pass

wky17 opened the pull request:
## [WIP][FIRRTL] Add formally verified alternative InferWidths pass

### 🔍 Related Issue
Fixes #9140 (InferWidths failures on cyclic constraints)

### 📖 Background & Motivation
The current `InferWidths` implementation has critical limitations:
**Correctness gaps**: Fails on cyclic constraint cases like [issue9140_0.mlir](https://github.com/wky17/circt/blob/issue-9140-new-inferwidths/test_new_inferwidths/issue9140_0.mlir) and [issue9140_1.mlir](https://github.com/wky17/circt/blob/issue-9140-new-inferwidths/test_new_inferwidths/issue9140_1.mlir)

This PR introduces `InferWidths_new` - a drop-in alternative pass that:
- Solves the correctness issues in #9140
- Comes with formal verification guarantees

### ⚙️ Technical Approach
The new pass implements a **constraint-based width inference engine** with:
- **Graph propagation core**: Using Boost's graph algorithms for: SCC decomposition (Tarjan)
- **Adapted Floyd-Warshall algorithm** and **Adapted Branch-and-Bound algorithm **: Solving inside SCC
- **Formal verification**: Core algorithm mechanically proven in [Rocq](https://rocq-prover.org)
- **Incremental adoption**: Preserves existing pass pipeline

> **Implementation note**: 
> Boost is used for rapid prototyping of graph algorithms. We have prepped replacement stubs for:
> - `Tarjan's algorithm` (SCC decomposition)
> - `Floyd-Warshall` (computing the least path)
> - `DFS` (finding any existing path between two nodes)
> These can be swapped in if necessary 

### 🧪 Testing Strategy
Comprehensive verification across 2 test dimensions:

| Test Category | Coverage | Verified Method |
|---------------|----------|---------------------|
| **Legacy compliance** | 100% existing tests<br>(`test/Dialect/FIRRTL/infer-widths.mlir`) | ✅ Passes all 60 existing tests |
| **Issue regression** | `test_new_inferwidths/`<br>• issue9140_0.mlir (example 1)<br>• issue9140_1.mlir (example 2)<br> | 🐛 Fixes #9140 failures<br> |

### 🚀 Integration Plan
Following maintainer's phased adoption strategy:
1. **[Current PR]** Add alternative pass with `firrtl-infer-widths-new` opt-in
2. **[Next]** Enable in downstream CI with shadow testing
3. **[Phase 3]** Performance benchmarking (memory/runtime)
4. **[Future]** Switch default after bake period

uenoku commented on lib/Dialect/FIRRTL/Transforms/InferWidths_new.cpp:
```diff
@@ -0,0 +1,3510 @@
+      .Default([](auto) { return false; });
+}
+
+using Valuation = std::unordered_map<FieldRef, int32_t, FieldRef::Hash>;
```
Can we use llvm::DenseMap instead of unorderd_map? By doing so it wouldn't be necessary to define Zero/Hash structs.

uenoku commented on lib/Dialect/FIRRTL/Transforms/InferWidths_new.cpp:
```diff
@@ -0,0 +1,3510 @@
+    } 
+};
+
+#include <list>
```
nit: Please include header at the top of file

darthscsi commented on lib/Firtool/Firtool.cpp:
```diff
@@ -90,7 +90,8 @@ LogicalResult firtool::populateCHIRRTLToLowFIRRTL(mlir::PassManager &pm,
 
   // Width inference creates canonicalization opportunities.
-  pm.nest<firrtl::CircuitOp>().addPass(firrtl::createInferWidths());
+  //pm.nest<firrtl::CircuitOp>().addPass(firrtl::createInferWidths());
```
Please either add a cmdline control over this or remove the old line.  Don't leave commented out code in the file.

darthscsi commented on lib/Dialect/FIRRTL/Transforms/InferWidths_new.cpp:
```diff
@@ -0,0 +1,3510 @@
+  if (mapping.areAllModulesSkipped())
+    return markAllAnalysesPreserved();
+
+  auto solve_time = std::chrono::high_resolution_clock::now();
```
See if MLIR's timing mechanisms are sufficient.

darthscsi commented on lib/Dialect/FIRRTL/Transforms/InferWidths_new.cpp:
```diff
@@ -0,0 +1,3510 @@
+    return success();
+
+  bool mappingFailed = false;
+  TypeSwitch<Operation *>(op)
```
Consider using an OpVisitor.

wky17 commented:
@schulyer Hi!  I have resolved the conflict in Firtool.cpp, please re-run again!

wky17 commented:
@seldridge @darthscsi Hi! I have resolved the conflict in Firtool.cpp, please re-run again!

wky17 commented:
@seldridge @darthscsi Dear reviewers,
All CI checks have passed after the recent fixes. Could you please review this PR when you have time?
Thank you for your time and consideration!

wky17 commented:
Hi @uenoku @seldridge 👋,  
Thanks for your thorough review! I've addressed all your suggestions in the latest push. Below is a breakdown of the changes:

### ✅ Implemented Suggestions:
1. **Include Statements**  
   - Moved `#include <list>` and `#include <stack>` to the top of the file for better organization.

2. **Memory Safety Improvement**  
   - Replaced unsafe `deque` usage with a `bump allocator` as suggested.

3. **Debugging Optimization**  
   - Wrapped entire loops in `LLVM_DEBUG` blocks.

4. **Test Pipeline Enhancement**  
   - Added explicit verification command to `infer-widths.mlir`:  
     ```bash
     circt-opt --pass-pipeline='builtin.module(firrtl.circuit(firrtl-infer-widths-new))' --verify-diagnostics %s | FileCheck %s
     ```  
   - **Verification:** All tests pass successfully ✅

5. **Code Documentation**  
   - Added detailed comments for:
     - `Constraint1`/`Constraint2`/`Min` classes
     - Key function behaviors throughout the file
   - *(Let me know if any section needs further clarification!)*

### ➡️ Next Steps:
- Happy to refine any remaining points
- Will address additional feedback
- Appreciate your expertise! 🙌

wky17 commented:
Hi @uenoku @seldridge 👋,

Thank you for your feedback on the PR. I've addressed the memory safety concerns by implementing the suggested `BumpPtrAllocator` approach. However, I'm encountering an unexpected CI failure that needs your help.

### CI Failure Analysis
The Windows CI test is failing with this error:
```plaintext
RUN: at line 2

d:\a\circt\circt\build\bin\circt-opt.exe --pass-pipeline='builtin.module(firrtl.circuit(firrtl-infer-widths-new))' --verify-diagnostics D:\a\circt\circt\test\Dialect\FIRRTL\infer-widths.mlir | d:\a\circt\circt\build\bin\filecheck.exe D:\a\circt\circt\test\Dialect\FIRRTL\infer-widths.mlir

executed command: 'd:\a\circt\circt\build\bin\circt-opt.exe' '--pass-pipeline=builtin.module(firrtl.circuit(firrtl-infer-widths-new))' --verify-diagnostics 'D:\a\circt\circt\test\Dialect\FIRRTL\infer-widths.mlir'

.---command stderr------------
D:\a\circt\circt\test\Dialect\FIRRTL\infer-widths.mlir:468:5: error: unexpected error: 'firrtl.domain.define' op not supported in width inference
    firrtl.domain.define %B, %A
    ^
```

### Investigation Done
I've thoroughly investigated this issue:

1. **File content verification**:
   - The referenced line (468) in `infer-widths.mlir` does **not** contain `firrtl.domain.define %B, %A`
   - The entire file `infer-widths.mlir` has **zero occurrence** of `firrtl.domain.define`

2. **Reproduction attempts**:
   - ✅ **Ubuntu 22.04**: `ninja -C build check-circt` passes all tests
   - ✅ **macOS (Sequoia 15.5)**: `ninja -C build check-circt` passes all tests
   - ❌ **Docker environment** (using `./utils/run-docker.sh`): cannot reproduce this error

### Request for Assistance
Since I can't reproduce this locally, could you please help with:
1. Confirming if this is a known Windows CI issue
2. Suggesting additional debugging steps
3. Recommending Windows-specific test commands I could try

I'm happy to investigate further with your guidance. The complete test file and changes can be seen in the PR.

seldridge commented:
It should be more than just Windows. See this commit: https://github.com/llvm/circt/commit/1be4b777dced6de52b9a31df779fb27f8ed0026a

Domain information can show up and it should ignored.

wky17 commented:
Hi @uenoku @seldridge 👋, I have now resolved the upstream conflicts and can merge cleanly. Please help me review the code again.

wky17 commented:
Hi @uenoku @seldridge 👋,

Thank you for catching the performance regression in the earlier version and sharing those critical metrics! 🙏 I've taken this feedback seriously and conducted optimization on the width inference algorithm.

### Performance Improvements
Using the largest available FIRRTL design I could access (309MB), here are the optimized results:

| Version          | `solver.solve()` Time | Speed vs Old |
|------------------|-----------------------|-------------|
| **Old (original)** | 49.337 ms             |  1.0x        |
| **Initial PR**     | 204.366 ms            | 4.14x slower  |
| **Optimized (the latest commitment)** | **11.123 ms**     | **4.44x faster** |

Key optimizations:
1. **Eliminated Heavy Copy Operations**
   Avoided copying large intermediate results (e.g., maps with hundreds of thousands of entries) during variable declarations and function result passing by using move semantics and pointers.

2. **Constraint Map Pruning**
   Proactively removed solved variables from the constraint map after resolution, reducing the search space for subsequent operations. This is safe due to our topological sorting approach and improves performance as the algorithm progresses.

Could you please re-run your 1.7GiB test case with this optimized implementation? I'm confident you'll see significant improvements.

uenoku commented on lib/Dialect/FIRRTL/Transforms/InferWidths_new.cpp:
```diff
@@ -0,0 +1,3240 @@
+                                 "path cannot be calculated."
+                              << "\n");
+    }
+  }
```
Please move this the LLVM_DEBUG below.
